ppafm/cpp/ProbeParticle.cpp

```diff
@@ -341,6 +345,118 @@ DLLEXPORT void getInPoints_LJ( int npoints, double * points_, double * FEs, int
     }
 }
 
+
+DLLEXPORT void computeD3Coeffs(
+    const int natoms_, const double *rs, const int *elems, const double *r_cov, const double *r_cut, const double *ref_cn,
+    const double *ref_c6, const double *r4r2, const double *k, const double *params, const int elem_pp, double *d3_coeffs
+) {
+
+    natoms = natoms_;
+    Ratoms = (Vec3d*)rs;
+
+    const double k1  = k[0];
+    const double k2  = k[1];
+    const double k3  = -k[2];
+    const double k12 = -k1 * k2;
+
+    const double s6 = params[0];
+    const double s8 = params[1];
+    const double a1 = params[2];
+    const double a2 = params[3];
+
+    float L[MAX_REF_CN * MAX_REF_CN];
+
+    // Compute reference C6 weights for the probe particle
+    int pp_ind = elem_pp - 1;
+    double L_pp[MAX_REF_CN];
+    int i;
+    for (i = 0; i < MAX_REF_CN; i++) {
+        double pp_cn = ref_cn[pp_ind * MAX_REF_CN + i];
+        if (pp_cn >= 0.0f) { // Values less that 0 zero are invalid and should not be counted.
+            L_pp[i] = exp(k3 * pp_cn * pp_cn);
+        }
+    }
+    int max_ref_pp = i;
+
+
+    for (int ia = 0; ia < natoms; ia++) { // Loop over all atoms
+
+        const Vec3d pos = Ratoms[ia];
+        const int elem_ind = elems[ia] - 1;
+        const double r_cov_elem = r_cov[elem_ind];
+
+        // Compute the coordination number for this atom
+        double cn = 0;
+        for (int j = 0; j < natoms; j++) {
+            if (j == ia) continue; // No self-interaction for coordination number
+            double d = (Ratoms[j] - pos).norm();
+            double r = r_cov[elems[j] - 1] + r_cov_elem;
+            cn += 1.0f / (1.0f + exp(k12 * r / d + k1));
+        }
+
+        // Compute gaussian weights and normalization factor for all of the reference
+        // coordination numbers.
+        double norm = 0;
+        int a, b;
+        for (a = 0; a < MAX_REF_CN; a++) {
+            double ref_cn_a = ref_cn[elem_ind * MAX_REF_CN + a];
+            if (ref_cn_a < 0.0f) break; // Invalid values after this
+            double diff_cn_a = ref_cn_a - cn;
+            double L_a = exp(k3 * diff_cn_a * diff_cn_a);
+            for (b = 0; b < max_ref_pp; b++) {
+                double L_ab = L_a * L_pp[b];
+                norm += L_ab;
+                L[a * MAX_REF_CN + b] = L_ab;
+            }
+        }
+        int max_ref_a = a;
+
+        // If the coordination number is so high that the gaussian weights are all zero,
+        // then we put all of the weight on the highest reference coordination number.
+        if (norm == 0) {
```

Comment:
Exactly zero ? I would think there should be some trashold as gaussians are never exactly zero. But probably I don't understand the algorithm?

Reply:
Yes, in principle a Gaussian is never zero, but due to finite precision, if the distance is big enough, then it will round to exactly zero, and you end up dividing by zero later, so I put this step in. It might not be a problem with double precision (didn't test), but it was definitely a problem with single precision in OpenCL, so I put it here as well.